File: Leetcode/2140 - Solving Questions With Brainpower/solution.js
var mostPoints = function(questions) {
    this.memoizedCalculations = {};
    this.questions = questions;
    return calcMaxPoints(0);
};

var calcMaxPoints = function(index) {
    if (index >= this.questions.length) {
        this.memoizedCalculations[index] = 0;
        return 0;
    }
    const points = this.questions[index][0];
    const brainpower = this.questions[index][1];
    if (this.memoizedCalculations[index] === undefined) {
        const pointsIfQuestionSolved = points + calcMaxPoints(index + brainpower + 1);
        const pointsIfQuestionNotSolved = calcMaxPoints(index + 1);
        this.memoizedCalculations[index] = Math.max(pointsIfQuestionSolved, pointsIfQuestionNotSolved);
    }
    return this.memoizedCalculations[index];
}

File: Leetcode/2140 - Solving Questions With Brainpower/solution.py
class Solution(object):
    def mostPoints(self, questions):
        memoized_calculations = {}
    
        def calc_max_points(index):
            if index >= len(questions):
                memoized_calculations[index] = 0
                return 0
            
            if index not in memoized_calculations:
                points, brainpower = questions[index]
                points_if_solved = points + calc_max_points(index + brainpower + 1)
                points_if_not_solved = calc_max_points(index + 1)
                memoized_calculations[index] = max(points_if_solved, points_if_not_solved)
            
            return memoized_calculations[index]
        
        return calc_max_points(0)
        